### PR TITLE
Implement svdx(Number, Number)

### PR DESCRIPTION
**Pull request by andreasnoack**
By popular demand. Also

Make `svd(x::Number)` return numbers

Implement `hash`, `isequal`, and `==` for `Factorization`s

cc @alanedelman

### CODE REVIEW

File: base/linalg/svd.jl [hunk 1]
```diff
@@ -231,6 +232,10 @@ function svdfact(A::StridedMatrix{TA}, B::StridedMatrix{TB}) where {TA,TB}
     S = promote_type(Float32, typeof(one(TA)/norm(one(TA))),TB)
     return svdfact!(copy_oftype(A, S), copy_oftype(B, S))
 end
+# This method can be heavily optimized but it is probably not cirtical
```
Comment (tkelman) [inline base/linalg/svd.jl]: critical

File: base/linalg/factorization.jl [hunk 2]
```diff
@@ -30,6 +30,10 @@ end
 convert(::Type{Factorization{T}}, F::Factorization{T}) where {T} = F
 inv(F::Factorization{T}) where {T} = A_ldiv_B!(F, eye(T, size(F,1)))
 
+Base.hash(F::Factorization, h::UInt) = mapreduce(f -> hash(getfield(F, f)), hash, h, fieldnames(F))
+Base.:(==)(  F::T, G::T) where {T<:Factorization} = all(f -> getfield(F, f) == getfield(G, f), fieldnames(F))
```
Comment (Sacha0) [inline base/linalg/factorization.jl]: Guessing the extra spaces are for alignment? :)

File: base/linalg/svd.jl [hunk 3]
```diff
@@ -231,6 +232,10 @@ function svdfact(A::StridedMatrix{TA}, B::StridedMatrix{TB}) where {TA,TB}
     S = promote_type(Float32, typeof(one(TA)/norm(one(TA))),TB)
     return svdfact!(copy_oftype(A, S), copy_oftype(B, S))
 end
+# This method can be heavily optimized but it is probably not critical
+# and might introduce bugs or inconsistencies relative to the 1x1 matrix
+# version
+svdfact(x::Number, y::Number) = svdfact(fill(x,1,1), fill(y,1,1))
```
Comment (Sacha0) [inline base/linalg/svd.jl]: Missing spaces between function arguments? (Spaces exist in the similar method definition below.)

File: test/linalg/svd.jl [hunk 4]
```diff
@@ -79,3 +79,25 @@ a2img  = randn(n,n)/2
...
+
+@testset "Number input" begin
+    x, y = randn(2)
+    @test svdfact(x)   == svdfact(fill(x,1,1))
+    @test svdvals(x)   == first(svdvals(fill(x,1,1)))
+    @test svd(x)       == first.(svd(fill(x,1,1)))
+    @test svdfact(x,y) == svdfact(fill(x,1,1), fill(y,1,1))
+    @test svdvals(x,y) == first(svdvals(fill(x,1,1), fill(y,1,1)))
+    @test svd(x,y)     == first.(svd(fill(x,1,1), fill(y,1,1)))
+end
```
Comment (Sacha0) [inline test/linalg/svd.jl]: We should choose a convention for spaces separating function arguments and use it consistently :).

**Comment by tkelman** [comment]
How many of the other factorizations do we do this for?

**Comment by andreasnoack** [comment]
In theory, we do all of them, see https://github.com/JuliaLang/julia/pull/1871, but some of them are broken, e.g. `qr`. I working on a PR to fix them.

**Pull request merged**